lib/puppet/parser/functions/fetch.rb

```diff
@@ -0,0 +1,70 @@
+module Puppet::Parser::Functions
+  newfunction(
+      :fetch,
+      :type => :rvalue,
+      :arity => -2,
+      :doc => <<-eos
+Looks up into a complex structure of arrays and hashes and returns a value
+or the default value if nothing was found.
+
+Key can contain slashes to describe path components. The function will go down
+the structure and try to extract the required value.
+
+$data = {
+  'a' => {
+    'b' => [
+      'b1',
+      'b2',
+      'b3',
+    ]
+  }
+}
+
+$value = fetch($data, 'a/b/2', 'not_found', '/')
+=> $value = 'b3'
+
+a -> first hash key
+b -> second hash key
+2 -> array index starting with 0
+
+not_found -> (optional) will be returned if there is no value or the path did not match. Defaults to nil.
+/ -> (optional) path delimiter. Defaults to '/'.
+
+In addition to the required "key" argument, "fetch" accepts default
+argument. It will be returned if no value was found or a path component is
+missing. And the fourth argument can set a variable path separator.
+  eos
+  ) do |args|
+
+    path_lookup = lambda do |data, path, default|
+      # no data -> return default
+      break default if data.nil?
+      # wrong path -> return default
+      break default unless path.is_a? Array
+      # empty path -> value found
+      break data unless path.any?
+      # non-empty path and non-structure data -> return default
+      break default unless data.is_a? Hash or data.is_a? Array
```

Comment:
All of these conditions create a complex beast. Having a debug output specifying what is being looked up and which decisions were made would greatly help users of this.